### PR TITLE
Add dirname, basename and extname getters

### PR DESCRIPTION
**Pull request by jeremyruppel**
Not sure if you guys would consider this bloat or a useful feature, but I often find myself wanting these pathname-style helpers when working with vinyl Files. Cheers!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -158,6 +158,36 @@ Object.defineProperty(File.prototype, 'relative', {
   }
 });
 
+Object.defineProperty(File.prototype, 'dirname', {
```
Comment (contra) [inline index.js]: file.base already exists, using dirname for output would be an antipattern

File: index.js [hunk 2]
```diff
@@ -158,6 +158,36 @@ Object.defineProperty(File.prototype, 'relative', {
...
+  get: function() {
+    if (!this.path) throw new Error('No path specified! Can not get basename.');
+    return path.basename(this.path);
+  },
+  set: function() {
+    throw new Error('File.basename is generated from the base and path attributes. Do not modify it.');
+  }
+});
+
+Object.defineProperty(File.prototype, 'extname', {
```
Comment (contra) [inline index.js]: file.extension already exists?

Comment (jeremyruppel) [inline index.js]: Does it? I can't find it anywhere in vinyl. Perhaps it was removed?

File: index.js (hunk 1 above)
Comment (jeremyruppel) [inline index.js]: dirname has a different meaning and can actually _be_ different than file.base. For example:

``` js
gulp.src('assets/css/**/*.scss')
```

Every file's file.base will be `assets/css` but any file in a subdirectory will have a different dirname.

File: index.js (hunk 2 above)
Comment (sindresorhus) [inline index.js]:
```js
console.log(file.extension);
```

inside a vinyl stream gives `undefined`.

**Comment by sindresorhus** [comment]
:+1: This would be very handy. I especially like being able to do `file.extension = '.foo' to change the extension, instead of having to use the gutil helper.

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: :+1: Yup, different use-case.

**Comment by contra** [comment]
@phated thoughts?

File: index.js [hunk 3]
```diff
@@ -158,6 +158,36 @@ Object.defineProperty(File.prototype, 'relative', {
...
+  get: function() {
+    if (!this.path) throw new Error('No path specified! Can not get extname.');
+    return path.extname(this.path);
+  },
+  set: function() {
+    throw new Error('File.extname is generated from the base and path attributes. Do not modify it.');
```
Comment (contra) [inline index.js]: i do agree that if we have a setter for this it should use the replace-ext module on path

**Comment by jeremyruppel** [comment]
Right on, I've added setters for all of these properties. What do you guys think?

File: index.js [hunk 4]
```diff
@@ -158,6 +159,36 @@ Object.defineProperty(File.prototype, 'relative', {
...
+  }
+});
+
+Object.defineProperty(File.prototype, 'extname', {
+  get: function() {
+    if (!this.path) throw new Error('No path specified! Can not get extname.');
+    return path.extname(this.path);
+  },
+  set: function(extname) {
+    this.path = replaceExt(this.path, extname);
```
Comment (contra) [inline index.js]: setters should also throw the errors on no path

Comment (jeremyruppel) [inline index.js]: I agree and personally prefer that they throw. Added.

**Comment by sindresorhus** [comment]
@jeremyruppel Can you also update the docs?

**Comment by jeremyruppel** [comment]
Just squashed these commits. @contra @sindresorhus @phated thoughts?

**Comment by contra** [comment]
@jeremyruppel Rebase?

**Comment by jeremyruppel** [comment]
@contra yep! README conflicts. Sorry about the trailing whitespace stuff.

**Comment by contra** [comment]
Thanks @jeremyruppel !

**Comment by jeremyruppel** [comment]
Thanks @contra & @sindresorhus!

File: README.md [hunk 5]
```diff
@@ -38,42 +38,42 @@ var coffeeFile = new File({
 
 #### options.cwd
 
-Type: `String`
```
Comment (stevemao) [inline README.md]: I don't think this whitespace should change.

Comment (tkellen) [inline README.md]: Markdown treats double trailing whitespace as a newline. It's stupid, but @stevemao is right.

Comment (jeremyruppel) [inline README.md]: Great point. #52

Comment (contra) [inline README.md]: TIL

**Pull request merged**